sycl/unittests/accessor/AccessorIterator.cpp

```diff
@@ -0,0 +1,323 @@
+#include <gtest/gtest.h>
+
+#include <sycl/sycl.hpp>
+
+#include <iterator>
+#include <numeric>
+#include <type_traits>
+#include <vector>
+
+class AccessorIteratorTest : public ::testing::Test {
+public:
+  AccessorIteratorTest() {}
+
+  template <int Dimensions, typename T = int>
+  void checkFullCopyThroughIterator(const sycl::range<Dimensions> &shape) {
+    std::vector<T> reference(shape.size());
+    std::iota(reference.begin(), reference.end(), 0);
+    sycl::buffer<T, Dimensions> buffer(reference.data(), shape);
+    auto accessor = buffer.template get_access<sycl::access_mode::read_write>();
+    std::vector<T> copied;
+    auto I = accessor.begin();
+    I = accessor.end();
+    for (auto i = accessor.begin(), e = accessor.end(); i != e; ++i) {
+      copied.push_back(*i);
+    }
+
+    ASSERT_EQ(copied.size(), reference.size());
+    for (size_t i = 0, e = reference.size(); i < e; ++i) {
+      ASSERT_EQ(copied[i], reference[i]);
+    }
+  }
+
+  template <int Dimensions, typename T = int>
+  void checkPartialCopyThroughIteratorWithoutOffset(
+      const sycl::range<Dimensions> &fullShape,
+      const sycl::range<Dimensions> &copyShape) {
+    std::vector<T> reference(fullShape.size());
+    std::iota(reference.begin(), reference.end(), 0);
+    sycl::buffer<T, Dimensions> buffer(reference.data(), fullShape);
+    std::vector<T> copied;
+    {
+      auto accessor =
+          buffer.template get_access<sycl::access_mode::read_write>(copyShape);
+      auto I = accessor.begin();
+      I = accessor.end();
+      for (auto i = accessor.begin(), e = accessor.end(); i != e; ++i) {
+        copied.push_back(*i);
+      }
+    }
+
+    ASSERT_EQ(copied.size(), copyShape.size());
+
+    {
+      auto fullAccessor = buffer.template get_access<sycl::access_mode::read>();
+
+      if constexpr (Dimensions == 1) {
```

Comment:
`dim_loop`

Reply:
Yes, I agree now. How about the following:

https://godbolt.org/z/738cPneKM for the idea, then

```
  range<3> copyShape3D = /* init lower_dims from copyShape */;
  for (...)
  for (...)
  for (...) {
    access<3 - Dims>(reference, i, j, k);
  }
```

Reply:
Thanks for the idea, implemented in 2dff81dc73ee29119e26430f50fe5daaf02b1f6f